spec/models/state_changers_spec.rb

```diff
@@ -70,122 +72,181 @@
+    let(:tube_failed_state) { 'failed' }
 
-        subject { StateChangers::TubeRackStateChanger.new(api, labware_uuid, user_uuid) }
+    let(:target_state) { 'qc_complete' }
 
-        before { stub_v2_tube_rack(tube_rack) }
+    let(:tube1_uuid) { SecureRandom.uuid }
+    let(:tube2_uuid) { SecureRandom.uuid }
+    let(:tube3_uuid) { SecureRandom.uuid }
 
-        context 'when all tubes are in failed state' do
-          let(:coordinates_to_pass) { [] }
+    let(:tube1) { create :v2_tube, uuid: tube1_uuid, state: tube_failed_state, barcode_number: 1 }
+    let(:tube2) { create :v2_tube, uuid: tube2_uuid, state: tube_starting_state, barcode_number: 2 }
+    let(:tube3) { create :v2_tube, uuid: tube3_uuid, state: tube_starting_state, barcode_number: 3 }
 
-          before { allow(labware).to receive(:racked_tubes).and_return([racked_tube1]) }
+    let!(:tube_rack) { create :tube_rack, barcode_number: 4, uuid: labware_uuid }
 
-          # if all the tubes are already in the target state expect contents to be empty
-          # TODO: I'm not sure this is correct behaviour, it should probably raise an error
-          # or a validation should catch that the state change is not needed
-          it 'returns empty array' do
-            expect(subject.contents_for(target_state)).to eq([])
-            subject.move_to!(target_state, reason, customer_accepts_responsibility)
-          end
-        end
+    let(:racked_tube1) { create :racked_tube, coordinate: 'A1', tube: tube1, tube_rack: tube_rack }
+    let(:racked_tube2) { create :racked_tube, coordinate: 'B1', tube: tube2, tube_rack: tube_rack }
+    let(:racked_tube3) { create :racked_tube, coordinate: 'C1', tube: tube3, tube_rack: tube_rack }
 
-        context 'when some tubes are not in failed state' do
-          let(:coordinates_to_pass) { %w[B1 C1] }
+    let(:labware) { tube_rack }
 
-          before { allow(labware).to receive(:racked_tubes).and_return([racked_tube1, racked_tube2, racked_tube3]) }
+    subject { StateChangers::TubeRackStateChanger.new(api, labware_uuid, user_uuid) }
 
-          it 'returns the coordinates of tubes not in failed state' do
-            expect(subject.contents_for(target_state)).to eq(%w[B1 C1])
-            subject.move_to!(target_state, reason, customer_accepts_responsibility)
-          end
-        end
+    before do
+      stub_v2_tube_rack(tube_rack)
+
+      # allow(labware).to receive(:racked_tubes).and_return([racked_tube1, racked_tube2, racked_tube3])
+    end
+
+    context 'when all tubes are in failed state' do
+      let(:coordinates_to_pass) { [] }
+
+      before do
+        # stub_v2_tube_rack(tube_rack)
+        allow(labware).to receive(:racked_tubes).and_return([racked_tube1])
       end
+
+      # if all the tubes are already in the target state expect contents to be empty
+      # TODO: I'm not sure this is correct behaviour, it should probably raise an error
+      # or a validation should catch that the state change is not needed
```

Comment:
I've seen this before in code on the sequencescape side. It gets an api call for a state change but then has code that works out nothing needs doing. Waste of an api call.
Not sure how that even gets triggered though. If there's a button that triggers a state change that isn't needed, then the button should not be visible (i.e. validation like you said).
Maybe someone decided it was too expensive to pull all the information into Limber and instead let SS decide?